text/0119-dynamic-component-creation.md

```diff
@@ -0,0 +1,335 @@
+
+## Detailed design
+
+A new property, `lightningElementName` on `LightningElementConstructor` is introduced to store the name of the custom element.
+
+A special tag, `<lwc.dynamic>` along with a new template directive `lwc:is` are also introduced and serves as the anchor in the DOM where the dynamic component will be rendered.
+
+Broken down, there are three main parts to this design:
+
+1. Storing and retrieving a custom element's name.
+2. Defining an anchor in the DOM.
+3. Instantiating the dynamic custom element.
+
+#### Storing and retrieving a custom element's name
+
+In today's world, an LWC module's namespace and name are known at compile time by [examining the directory structure](https://rfcs.lwc.dev/rfcs/lwc/0101-module-unification) and through [module resolution](https://rfcs.lwc.dev/rfcs/lwc/0020-module-resolution).
+
+The default folder structure for an LWC module looks as follows:
+
+```
+namespace/
+└── name/
+    ├── *.css
+    ├── *.js
+    ├── *.html
+```
+
+Leveraging this, the namespace and name can be accessed at compile time and ultimately stored on the constructor.  A new property on `LightningElementConstructor` called `lightningElementName` will store the custom element name which will be of the form `namespace-name`.
+
+```javascript
+const { default: ctor } = await import("c/customConstructor");
+// The tag name can be directly accessed on the constructor
+ctor.lightningElementName
+```
+
+The value of `lightningElementName` should only be set by the LWC engine to prevent component authors from modifying it and producing unintended side effects.
+
+Alternatively, the custom element name can be stored in a `weakMap` and a public API can be exposed to retrieve the constructor's custom element name. However, because the custom element name can be used as a selector with `this.template.querySelector` placing it on the constructor seems more ergonomic.
+
+#### Defining an anchor in the DOM
+
+##### `<lwc.dynamic>` tag
+
+`<lwc.dynamic>` is a placeholder tag for the dynamic component and will not be rendered to the DOM.  It serves as a special signal to the compiler that the dynamic component will be rendered at the specific location in the DOM. (Similar to how `template` is not rendered)
+
+Note the separation of the namespace and name using a '.' is intentional because of the following reasons:
+
+1. It is neither a valid [tag name](https://html.spec.whatwg.org/multipage/syntax.html#syntax-tag-name) nor a valid [custom element name](https://html.spec.whatwg.org/multipage/custom-elements.html#valid-custom-element-name) which should clearly signal that this is a special element (which will not be rendered to the DOM).
+2. Initially considered using a ':' instead of a '.' but ':' is the delimiter Aura uses to separate namespace and name.  To prevent confusion on whether this is an Aura tag, decided to use '.' instead.
```

Comment:
My 2 cents: While I understand the point about confusion with Aura components, in my opinion, it would be more idiomatic to use a `:` instead of a `.` here as it maps well to the concept of XML namespace. Since we would only permit `<lwc:dynamic>` for now, and maybe other `<lwc:*>` in the future, the compiler could log errors and warning messages when this syntax is misused.

Reply:
I also find it interesting that GitHub's syntax highlighting breaks when using the `.`:

```html
<lwc.dynamic></lwc.dynamic>
```

Purely for readability, I agree with @pmdartus about using a `:` character:

```html
<lwc:dynamic></lwc:dynamic>
```

Reply:
Sounds good, I've updated to use `lwc:dynamic`.